### PR TITLE
Fix routing table names for HA

### PR DESCRIPTION
**Pull request by wtty-fool**
Towards https://github.com/giantswarm/giantswarm/issues/12940
Related PR https://github.com/giantswarm/giantnetes-terraform/pull/380

Improves routing table names for:
- [x] tccp
- [x] tcnp

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Comment by xh3b4sd** [comment]
Just one thing, when we change the templates and its resources we need to make sure that upgrades work fine and that there is as little workload disruption as possible. Cloud Formation might be bitchy about renaming certain resources because it treats it like a replacement. So it destroys the world and creates it from the ashes.

**Comment by wtty-fool** [comment]
@xh3b4sd I've tested the update and it looked fine to my untrained eye. 
Since AWS uses `logical-id` to update CF resources, it did not mind the change of names.

Here are the updated route tables:
image

And here's proof they were updated in the process:
image

Let me know if my reasoning is off or I need to do more digging to ensure safe updates!

**Review by xh3b4sd — APPROVED** [review]
Considering the upgrades are fine it should be good to merge. Just for awareness I would like @calvix to acknowledge this change.

**Comment by wtty-fool** [comment]
FYI the final shape of RouteTables section:
image

File: service/controller/key/common.go
Comment (calvix) [inline service/controller/key/common.go:77]: this could use little bit of comment with an example, I had to read it several times and know what values is going in to figure out what it will return

**Review by calvix — APPROVED** [review]

**Comment by calvix** [comment]
LGTM if this was tested also with multiple node pools and updates

**Comment by wtty-fool** [comment]
It works with multiple node pools and updates :) 
However @corest pointed out we'd have to list all subnets in `RouteTables` flag now. We are currently working on improving this before merging this PR.

File: CHANGELOG.md
Comment (kopiczko) [inline CHANGELOG.md:21]: There is a `### Deprecated` section. We can add a note about `RouteTables` flag there.

Till we don't break the compatibility I think we can add this entry to `### Added` and once we remove `RouteTables` it will get added to `### Removed` and that will be the breaking change.

Headers.

File: service/controller/resource/cproutetables/resource.go
Comment (kopiczko) [inline service/controller/resource/cproutetables/resource.go:47]: Is it a TODO still?

File: service/controller/resource/cproutetables/resource.go
Comment (kopiczko) [inline service/controller/resource/cproutetables/resource.go:72]: Shouldn't this log message stay?

File: service/controller/resource/cproutetables/resource.go
Comment (kopiczko) [inline service/controller/resource/cproutetables/resource.go:111]: Does it make sense to have that `r.expectedTableCount`? I feel like it's only testing if `len` returns the same value for the same slice.

**Comment by wtty-fool** [comment]
I updated RouteTable caching. The idea is as follows:
**Short term**
When there are `names` supplied via `RouteTables` flag (which is what we do currently), we use them to look up RouteTables just like before. The new part is `lookupByTag`, which works when no `names` are given. It uses tags already existing in TC RouteTables to find all matching RouteTables. This way `aws-operator` works both on 1CP and 2CP out of the box.

**Long term**
We add a missing tag (`giantswarm.io/route-table-type`) to control planes https://github.com/giantswarm/giantnetes-terraform/pull/380. When all CPs are eventually rolled, and all route tables are tagged, we can drop `RouteTableNames` parameter from installations repo https://github.com/giantswarm/installations/blob/master/ginger/draughtsman-configmap-values.yaml#L39 and `routeTables` from configmap https://github.com/giantswarm/aws-operator/blob/master/helm/aws-operator/templates/configmap.yaml#L28

File: service/controller/resource/cproutetables/resource.go
Comment (wtty-fool) [reply service/controller/resource/cproutetables/resource.go:111]: RouteTables don't change over time. If we get rid of `r.expectedTableCount`, we are hitting AWS API every time to get the same list of tables. This way we ask once and perform lookup again if something changes.

Comment (kopiczko) [reply service/controller/resource/cproutetables/resource.go:111]: I'm asking if we can replace:

```
if len(r.routeTables) == r.expectedTableCount && r.expectedTableCount > 0 {
```

with:

```
if len(r.routeTables) > 0 {
```

and get rid of `r.expectedTableCount`.

Comment (wtty-fool) [reply service/controller/resource/cproutetables/resource.go]: Smart, I'll do that.

**Review by kopiczko — APPROVED** [review]
LGTM

**Comment by wtty-fool** [comment]
@xh3b4sd Pinging you in case you want to take another look before I merge :)

**Review by xh3b4sd — APPROVED** [review]
I think that in case your collaboration ensures that the cluster transitions work fine this is good to go. Thanks a lot everyone.

**Pull request merged**
